src/QuokkaSimulation.hpp

```diff
@@ -1643,20 +1703,26 @@ auto QuokkaSimulation<problem_t>::advanceHydroAtLevel(amrex::MultiFab &state_old
 		}
 
 		for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
-			amrex::MultiFab::Saxpy(flux_rk2[idim], 0.5, fluxArrays[idim], 0, 0, ncompHydro_, 0);
+			amrex::MultiFab::Saxpy(flux_rk2[idim], 0.5, fluxArrays[idim], 0, 0, nvars_, 0);
 			amrex::MultiFab::Saxpy(avgFaceVel[idim], 0.5, faceVel[idim], 0, 0, 1, 0);
 			if constexpr (Physics_Traits<problem_t>::is_mhd_enabled) {
 				amrex::MultiFab::Saxpy(ec_emf_components_rk_ave[idim], 0.5, ec_emf_components_rk_stage2[idim], 0, 0, 1, 0);
 			}
 		}
 
-		amrex::MultiFab rhs(grids[lev], dmap[lev], ncompHydro_, 0);
+		amrex::MultiFab rhs(grids[lev], dmap[lev], nvars_, 0);
+		amrex::MultiFab FOrhs(grids[lev], dmap[lev], nvars_, 0);
 		amrex::iMultiFab redoFlag(grids[lev], dmap[lev], 1, 1);
 		redoFlag.setVal(quokka::redoFlag::none);
 
-		HydroSystem<problem_t>::ComputeRhsFromFluxes(rhs, flux_rk2, dx, ncompHydro_);
-		HydroSystem<problem_t>::AddInternalEnergyPdV(rhs, stateOld_cc, dx, avgFaceVel, redoFlag);
-		HydroSystem<problem_t>::PredictStep(stateOld_cc, stateFinal_cc, rhs, dt_lev, ncompHydro_, redoFlag);
+		HydroSystem<problem_t>::ComputeRhsFromFluxes(rhs, fluxArrays, dx, nvars_);
+		HydroSystem<problem_t>::AddInternalEnergyPdV(rhs, stateOld_cc, dx, faceVel, redoFlag);
+
+		// calculate rhs = 0.5 * (state_inter - state_old) / dt_lev + 0.5 * rhs
+		rhs.mult(0.5, 0, nvars_, 0);
+		amrex::MultiFab::Saxpy(rhs, 0.5 / dt_lev, stateInter_cc, 0, 0, nvars_, 0);
+		amrex::MultiFab::Saxpy(rhs, -0.5 / dt_lev, stateOld_cc, 0, 0, nvars_, 0);
+		HydroSystem<problem_t>::PredictStep(stateOld_cc, stateFinal_cc, rhs, dt_lev, nvars_, redoFlag);
```

Comment:
The final state cannot be computed this way in the hydro update. The rhs must be computed from the sum of the stage 1 and stage 2 fluxes to maintain conservation when FOFC is activated.

It may or may not be possible to generalize this to include frictional source terms in the RK update, but regardless of the method, the hydro integrator needs to remain identical when dust is turned off.

Reply:
I'm guessing the dust damping test causes the gas velocity to increase, so the CFL stable timestep is reduced a lot in the first timestep. You should be able to make the first timestep small enough so that the retries aren't triggered.

Reply:
For testing purposes, you can have the hydro update always return `true`, which means it won't trigger the retries under any circumstances. But for production simulations, I didn't want to provide a way to disable that, because it will almost always lead to the simulation becoming unstable or crashing.

Reply:
It will be some time before I have time to review the new RK2 update in detail. I'll let you know.